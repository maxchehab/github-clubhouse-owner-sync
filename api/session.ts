import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../common/util/with-session.util';

import lambda from '../common/util/lambda.util';
import Session from '../common/interfaces/session.interface';

export default lambda(
  withSession(
    async (_req: NextApiRequest, res: NextApiResponse, session: Session) => {
      delete session.token;

      return res.json(session);
    },
  ),
);
