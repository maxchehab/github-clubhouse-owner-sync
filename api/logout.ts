import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../common/util/with-session.util';

import lambda from '../common/util/lambda.util';
import Session from '../common/interfaces/session.interface';
import { deleteSessionByID } from '../services/fauna/services/session.service';

export default lambda(
  withSession(
    async (_req: NextApiRequest, res: NextApiResponse, session: Session) => {
      await deleteSessionByID(session.id);

      res.writeHead(302, {
        'set-cookie': `sid=; Expires=${new Date(0).toUTCString()};`,
        'location': '/login',
      });
      res.end();
    },
  ),
);
