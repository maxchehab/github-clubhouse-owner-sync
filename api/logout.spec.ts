import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import * as SessionService from '../services/fauna/services/session.service';
import closeServer from '../test/close-server';
import createTestLambda from '../test/create-test-lambda';
import logout from './logout';

describe('api/logout', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(logout);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  describe('with GET api/logout', () => {
    let deleteSessionByID: jest.MockInstance<Promise<any>, [string]>;

    beforeEach(() => {
      deleteSessionByID = jest
        .spyOn(SessionService, 'deleteSessionByID')
        .mockImplementationOnce(async () => null);

      jest
        .spyOn(SessionService, 'findSessionByID')
        .mockImplementationOnce(async () => ({
          name: 'John Smith',
          expires_at: new Date().getTime() + 1000,
          token: 'access-token',
          id: 'sess_01DY49R4RBSET4H39RX7JKR0AG',
          avatar: 'https://example.com/avatar.png',
        }));
    });

    it('deletes session', async () => {
      const { status, headers } = await axios.get(url, {
        headers: {
          cookie: 'sid=sess_01DY49R4RBSET4H39RX7JKR0AG',
        },
        validateStatus: () => true,
        maxRedirects: 0,
      });

      expect(deleteSessionByID).toBeCalledTimes(1);
      expect(status).toEqual(302);
      expect(headers.location).toEqual('/login');
      expect(headers['set-cookie']).toMatchSnapshot();
    });
  });
});
