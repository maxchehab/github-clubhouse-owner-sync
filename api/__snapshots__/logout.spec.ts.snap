// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/logout with GET api/logout deletes session 1`] = `
Array [
  "sid=; Expires=Thu, 01 Jan 1970 00:00:00 GMT;",
]
`;
